tools/floodgate/js/promote.js

```diff
@@ -0,0 +1,167 @@
+import {
+  createFolder,
+  getAuthorizedRequestOption,
+  saveFile,
+  updateExcelTable,
+  validateConnection,
+} from '../../loc/sharepoint.js';
+import {
+  hideButtons,
+  loadingON,
+  showButtons,
+  simulatePreview,
+} from '../../loc/utils.js';
+import { getConfig as getFloodgateConfig } from './config.js';
+import { ACTION_BUTTON_IDS } from './ui.js';
+import { getFile, handleExtension } from './utils.js';
+
+/**
+ * Copies the Floodgated files back to the main content tree.
+ * Creates intermediate folders if needed.
+ */
+async function promoteCopy(srcPath, destinationFolder) {
+  validateConnection();
+  await createFolder(destinationFolder);
+  const { sp } = await getFloodgateConfig();
+  const destRootFolder = `${sp.api.file.copy.baseURI}`.split('/').pop();
+
+  const payload = { ...sp.api.file.copy.payload, parentReference: { path: `${destRootFolder}${destinationFolder}` } };
+  const options = getAuthorizedRequestOption({
+    method: sp.api.file.copy.method,
+    body: JSON.stringify(payload),
+  });
+
+  // copy source is the pink directory for promote
+  const copyStatusInfo = await fetch(`${sp.api.file.copy.fgBaseURI}${srcPath}:/copy`, options);
+  const statusUrl = copyStatusInfo.headers.get('Location');
+  let copySuccess = false;
+  let copyStatusJson = {};
+  while (statusUrl && !copySuccess && copyStatusJson.status !== 'failed') {
```

Comment:
Why is this a loop?  Seems like an if statement would do?

Reply:
@chrischrischris When the copy is triggered using Sharepoint's copy API, it provides a status endpoint (as seen on line 36) that we need to check on to understand the outcome of the copy operation. I am assuming that's because they are handling the copy asynchronously on their end and cannot provide immediate response on the copy outcome. The status sometimes might take few milliseconds longer, so we loop and wait until it has some sort of outcome available. Hope that provides some context.